File: solutions/java/code-carl/ch07_backtracking/T8_LC0078_Subsets.java
/**
 * Practice template for:
 * - LeetCode 78. Subsets
 *   https://leetcode.cn/problems/subsets/
 */
import java.util.ArrayList;
import java.util.List;

public class T8_LC0078_Subsets {
    List<Integer> path = new ArrayList<>();
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtracking(nums, 0);
        return res;
    }

    public void backtracking(int[] nums, int startIdx) {
        res.add(new ArrayList<>(path));
        if (startIdx == nums.length) {
            return;
        }
        for (int i = startIdx; i < nums.length; i++) {
            path.add(nums[i]);
            backtracking(nums, i + 1);
            path.remove(path.size() - 1);
        }

    }
}

File: solutions/python/code-carl/ch07_backtracking/t8_lc0078_subsets.py
"""LeetCode 78. Subsets
https://leetcode.cn/problems/subsets/
"""

from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        path: List[int] = []
        ans: List[List[int]] = []

        def backtracking(nums: List[int], start_idx: int) -> None:
            ans.append(path[:])
            if start_idx == len(nums):
                return
            for i in range(start_idx, len(nums)):
                path.append(nums[i])
                backtracking(nums, i + 1)
                path.pop()

        backtracking(nums, 0)
        return ans


if __name__ == "__main__":
    solution = Solution()
    result = {tuple(x) for x in solution.subsets([1, 2, 3])}
    expected = {
        (),
        (1,),
        (2,),
        (3,),
        (1, 2),
        (1, 3),
        (2, 3),
        (1, 2, 3),
    }
    assert result == expected

    solution = Solution()
    result = {tuple(x) for x in solution.subsets([0])}
    assert result == {(), (0,)}

    print("All tests passed.")
